tutorials/data-science-preprocessing/index.md

```diff
@@ -0,0 +1,437 @@
+                    if line.startswith('}'):
+                        yield json.loads(''.join(content))
+                        break
+# end JsonFileSource
+```
+
+This `Source` looks for the beginning of a json object by searcing for a
+top-level open-curly-brace (`{`), and collects all subsequent lines until it
+finds a top-level close-curly-brace (`}`), whereupon it creates a json object
+from the accumulated string and emits it. The result is an iterator that emits
+json objects one at a time.
+
+Note that the method makes use of a [`RangeTracker`][rangetracker] to enable
+parallelization of the `Source`. That is, for large datasets, several workers
+may access different ranges in the same source file to process the data in
+parallel, to boost performance. The `Source`, then, is only responsible for
+emitting the json objects that begin within the range assigned to it, and should
+stop once it reaches the end of its range.
+
+[beam]: http://beam.incubator.apache.org/
+[filebasedsource]: /dataflow/model/custom-io-python#convenience-source-base-classes
+[rangetracker]: /dataflow/model/custom-io-python#implementing-the-rangetracker-subclass
+
+### Attach the functions in series
+
+Now that you've defined the source of data, you can now pipe it through to the
+filtering functions, in series. Apache Beam uses the pipe operator (`|`) to do
+this (with the double angle bracket operator (`>>`) to add a description):
+
+[embedmd]:# (src/preprocessing/clean.py /def main/ / p\.run.*/)
+```py
+def main(src_path, dest_table, pipeline_args):
+    p = apache_beam.Pipeline(argv=pipeline_args)
+
+    value = p | 'Read JSON' >> apache_beam.Read(JsonFileSource(src_path))
+
+    value |= (
+        'Remove records that lack location or year data' >>
+        apache_beam.FlatMap(discard_incomplete))
+
+    value |= (
+        'Convert string values to their types' >>
+        apache_beam.Map(convert_types))
+
+    value |= (
+        'Filter bad data' >>
+        apache_beam.FlatMap(filter_suspicious))
+
+    value |= (
+        'Massage fields with "rec" prefix' >>
+        apache_beam.Map(massage_rec))
+
+    value |= (
+        'Dump data to BigQuery' >>
+        apache_beam.Write(apache_beam.io.BigQuerySink(
+            dest_table,
+            schema=', '.join([
+                'fall:STRING',
+                'year:INTEGER',
+                'nametype:STRING',
+                'mass:FLOAT',
+                'name:STRING',
+                'class:STRING',
+                'latitude:FLOAT',
+                'longitude:FLOAT',
+                'id:STRING',
+            ]),
+            create_disposition=(
+                apache_beam.io.BigQueryDisposition.CREATE_IF_NEEDED),
+            write_disposition=(
+                apache_beam.io.BigQueryDisposition.WRITE_TRUNCATE))))
+
+    p.run()
+```
+
+Note that this example takes advantage of the built-in [`BigQuerySink`][bq-sink]
+to output the result into BigQuery, to analyze later.
```

Comment:
"..to output ..., to analyze" construction is a bit awkward. Maybe say something like, "..., so that you can analyze it later," or, "..., where you can..."